src/Farmer/Builders/Builders.HostGroup.fs

```diff
@@ -0,0 +1,132 @@
+﻿[<AutoOpen>]
+module Farmer.Builders.HostGroups
+
+open System
+open Farmer
+open Farmer.Arm
+open Farmer.DedicatedHosts
+    
+type HostGroupConfig =
+    {
+        Name: ResourceName
+        AvailabilityZones: string list option
+        SupportAutomaticPlacement: FeatureFlag option
+        UltraSSDEnabled: FeatureFlag option
+        PlatformFaultDomainCount: int option
+        Tags: Map<string, string>
+    }
+    interface IBuilder with
+        member this.ResourceId = hostGroups.resourceId this.Name
+        member this.BuildResources location =
+            let hostGroup: HostGroup =
+                {
+                    Name = this.Name
+                    Location = location
+                    AvailabilityZones = this.AvailabilityZones |> (Option.defaultValue List.Empty)
+                    SupportAutomaticPlacement = this.SupportAutomaticPlacement |> (Option.defaultValue FeatureFlag.Disabled)
+                    UltraSSDEnabled = this.UltraSSDEnabled |> (Option.defaultValue FeatureFlag.Disabled)
+                    PlatformFaultDomainCount = this.PlatformFaultDomainCount |> (Option.defaultValue 0)
+                    Tags = this.Tags
+                }
+            [ hostGroup ]
+
+type HostGroupBuilder() =
+    member _.Yield _ =
+        {
+            Name = ResourceName.Empty
+            AvailabilityZones = None
+            SupportAutomaticPlacement = None
+            UltraSSDEnabled = None
+            PlatformFaultDomainCount = None
+            Tags = Map.empty
+        }
+    [<CustomOperation "name">]
+    member _.Name(state: HostGroupConfig, name: string) = { state with Name = ResourceName name }
+    [<CustomOperation "add_availabilityZones">] // todo what do these even look like?
+    member _.AddAvailabilityZone(state: HostGroupConfig, az: string list) =
+        match state.AvailabilityZones with
+        | None -> { state with AvailabilityZones = Some az }
+        | Some curZones -> { state with AvailabilityZones = Some (az @ curZones) }
+    [<CustomOperation "supportAutomaticPlacement">]
+    member _.SupportAutomaticPlacement(state: HostGroupConfig, flag: FeatureFlag) =
+        { state with SupportAutomaticPlacement = Some flag }
+    [<CustomOperation "enableUltraSsd">]
+    member _.EnableUltraSSD(state: HostGroupConfig, flag: FeatureFlag) =
+        { state with UltraSSDEnabled = Some flag }
+    [<CustomOperation "platformFaultDomainCount">]
+    member _.PlatformFaultDomainCount(state: HostGroupConfig, domainCount: int) =
+        { state with PlatformFaultDomainCount = Some domainCount }
+        
+let hostGroup = HostGroupBuilder()
+    
+type HostConfig =
+    {
+        Name: ResourceName
+        AutoReplaceOnFailure: FeatureFlag option
+        LicenseType: HostLicenseType option
+        HostSku: HostSku option
+        PlatformFaultDomain: int option
+        HostGroupName: LinkedResource
+        PublicKey: Uri option
```

Comment:
this is a restricted feature so not sure how to handle that, it should be an xml public key